packages/caravan-fees/src/rbf.ts

```diff
@@ -0,0 +1,530 @@
+ *   originalTx: originalTxHex,
+ *   availableInputs: availableUTXOs,
+ *   cancelAddress: 'bc1q...', // cancel address
+ *   network: Network.MAINNET,
+ *   dustThreshold: '546',
+ *   scriptType: 'P2WSH',
+ *   requiredSigners: 2,
+ *   totalSigners: 3,
+ *   targetFeeRate: '10',
+ *   absoluteFee: '1000',
+ *   fullRBF: false,
+ *   strict: true
+ *   reuseAllInputs: false, // Default behavior, more efficient fee wise
+ * });
+ *
+ * @see https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki BIP 125: Opt-in Full Replace-by-Fee Signaling
+ * @see https://developer.bitcoin.org/devguide/transactions.html#locktime-and-sequence-number Bitcoin Core's guide on locktime and sequence numbers
+ * @see https://bitcoinops.org/en/newsletters/2023/10/25/#fn:rbf-warning Bitcoin Optech on replacement cycle attacks
+ */
+export const createCancelRbfTransaction = (
+  options: CancelRbfOptions,
+): string => {
+  const {
+    fullRBF = false,
+    strict = false,
+    originalTx,
+    network,
+    targetFeeRate,
+    absoluteFee,
+    availableInputs,
+    requiredSigners,
+    totalSigners,
+    scriptType,
+    dustThreshold,
+    cancelAddress,
+    reuseAllInputs = false,
+  } = options;
+  // Step 1: Analyze the original transaction
+  const txAnalyzer = new TransactionAnalyzer({
+    txHex: originalTx,
+    network,
+    targetFeeRate,
+    absoluteFee,
+    availableUtxos: availableInputs,
+    requiredSigners,
+    totalSigners,
+    addressType: scriptType,
+  });
+
+  txAnalyzer.assumeRBF = fullRBF;
+
+  const analysis = txAnalyzer.analyze();
+  const originalTxFee = new BigNumber(txAnalyzer.fee);
+
+  // Step 2: Verify RBF possibility and suitability
+  if (!analysis.canRBF) {
+    if (fullRBF) {
+      console.warn(
+        "Transaction does not signal RBF. Proceeding with full RBF, which may not be accepted by all nodes.",
+      );
+    } else {
+      throw new Error(
+        "RBF is not possible for this transaction. Ensure at least one input has a sequence number < 0xfffffffe.",
+      );
+    }
+  }
+
+  if (analysis.recommendedStrategy !== FeeBumpStrategy.RBF) {
+    console.warn(
+      "RBF is not the recommended strategy for this transaction. Consider using the recommended strategy: " +
+        analysis.recommendedStrategy,
+    );
+  }
+
+  // Step 3: Create a new transaction template
+  const newTxTemplate = new BtcTransactionTemplate({
+    inputs: [],
+    outputs: [],
+    targetFeeRate: Number(txAnalyzer.rbfFeeRate), // Use RBF-specific fee rate from analyzer
+    dustThreshold,
+    network,
+    scriptType,
+    requiredSigners,
+    totalSigners,
+  });
+
+  // Step 4: Add the cancellation output and delete all previous outputs
+
+  newTxTemplate.addOutput(
+    new BtcTxOutputTemplate({
+      address: cancelAddress,
+      amountSats: "0", // Temporary amount, will be adjusted later
+      locked: false,
+    }),
+  );
+
+  // Step 5: Add inputs from the original transaction
```

Comment:
Aesthetic suggestion: do you think it would make sense (or even work), to combine steps 5 and 6 into an internal function like `_addRBfInputs`. Then you might be able to reuse it for both accelerate and cancel and it would make these functions more readable.

Reply:
changes [done](https://github.com/caravan-bitcoin/caravan/pull/114/commits/c34de8a8ff6115cd401f7d5cb16406c991140064#diff-47122bb461c82a41c84d8866adf0cb005bf7f573a2afb8558f670a419bed146aR84)